### PR TITLE
[MangoTV] Improve extractor

### PR DESCRIPTION
**Pull request by hatienl0i261299**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Improve extractor for [MangoTV](https://w.mgtv.com/) and update testcase for zingmp3

### CODE REVIEW

File: yt_dlp/extractor/mgtv.py [hunk 1]
```diff
@@ -97,4 +129,23 @@ def _real_extract(self, url):
             'description': info.get('desc'),
             'duration': int_or_none(info.get('duration')),
             'thumbnail': info.get('thumb'),
+            'subtitles': self.get_all_subtitles(video_id=video_id, domain=stream_domain),
```
Comment (Ashish0804) [inline yt_dlp/extractor/mgtv.py:132]:
```suggestion
            'subtitles': self.get_all_subtitles(video_id, stream_domain),
```
Much cleaner.

File: yt_dlp/extractor/mgtv.py [hunk 2]
```diff
@@ -97,4 +129,23 @@ def _real_extract(self, url):
...
         }
+
+    def get_all_subtitles(self, video_id, domain):
+        info = self._download_json(f'https://pcweb.api.mgtv.com/video/title?videoId={video_id}', video_id=video_id)
```
Comment (Ashish0804) [inline yt_dlp/extractor/mgtv.py:136]:
```suggestion
        info = self._download_json(f'https://pcweb.api.mgtv.com/video/title?videoId={video_id}', video_id, fatal=False) or {}
```

File: yt_dlp/extractor/mgtv.py [hunk 3]
```diff
@@ -97,4 +129,23 @@ def _real_extract(self, url):
...
+        subtitles = {}
+        for sub in try_get(info, lambda x: x['data']['title']):
+            url_sub = sub.get('url')
+            if not url_sub:
+                continue
+            locale = sub.get('captionCountrySimpleName').lower()
+            sub = self._download_json(f'{domain}{url_sub}', video_id=video_id,
+                                      note=f'Download subtitle for locale {sub.get("name")} ({locale})')['info']
```
Comment (Ashish0804) [inline yt_dlp/extractor/mgtv.py:144]:
```suggestion
            sub = self._download_json(f'{domain}{url_sub}', video_id,
                                      note=f'Download subtitle for locale {sub.get("name")} ({locale})',
                                      fatal=False)
```

File: yt_dlp/extractor/mgtv.py [hunk 4]
```diff
@@ -97,4 +129,23 @@ def _real_extract(self, url):
...
+            if not sub:
+                continue
```
Comment (Ashish0804) [inline yt_dlp/extractor/mgtv.py:146]:
```suggestion
            sub_url = url_or_none((sub or {}).get('info'))
            if not sub_url:
                continue
```

File: yt_dlp/extractor/mgtv.py [hunk 5]
```diff
@@ -97,4 +129,23 @@ def _real_extract(self, url):
...
+            subtitles[locale] = [{
+                'url': url_or_none(sub),
+                'ext': 'srt'
+            }]
```
Comment (Ashish0804) [inline yt_dlp/extractor/mgtv.py:150]:
```suggestion
           subtitles.setdefault(locale or 'en', []).append({
                                'url': sub_url,
                                'ext': 'srt'
            })
```

File: yt_dlp/extractor/mgtv.py [hunk 6]
```diff
@@ -97,4 +129,23 @@ def _real_extract(self, url):
...
         }
+
+    def get_all_subtitles(self, video_id, domain):
+        info = self._download_json(f'https://pcweb.api.mgtv.com/video/title?videoId={video_id}', video_id=video_id)
+        subtitles = {}
+        for sub in try_get(info, lambda x: x['data']['title']):
+            url_sub = sub.get('url')
+            if not url_sub:
+                continue
+            locale = sub.get('captionCountrySimpleName').lower()
```
Comment (Ashish0804) [inline yt_dlp/extractor/mgtv.py:142]:
```suggestion
            locale = sub.get('captionCountrySimpleName')
```

**Comment by Ashish0804** [comment]
I'm aware that suggestions are not in proper order so read them all before making changes.
Also formatting is fucked.
Core idea here is: `Don't make sub extraction fatal.`

File: yt_dlp/extractor/zingmp3.py [hunk 7]
```diff
@@ -158,6 +158,8 @@ class ZingMp3IE(ZingMp3BaseIE):
             'duration': 207,
             'track': 'Sương Hoa Đưa Lối',
             'artist': 'K-ICM, RYO',
+            'album': 'Sương Hoa Đưa Lối (Single)',
+            'album_artist': 'K-ICM, RYO',
```
Comment (pukkandan) [inline yt_dlp/extractor/zingmp3.py:162]: remove from this PR. I'll commit it separately

File: yt_dlp/extractor/mgtv.py [hunk 8]
```diff
@@ -97,4 +129,26 @@ def _real_extract(self, url):
             'description': info.get('desc'),
             'duration': int_or_none(info.get('duration')),
             'thumbnail': info.get('thumb'),
+            'subtitles': self.get_all_subtitles(video_id, stream_domain),
         }
+
+    def get_all_subtitles(self, video_id, domain):
```
[the remark this answers is not in the document]
Comment (pukkandan) [reply yt_dlp/extractor/mgtv.py]: This was not a typo. `extract_subtitles` is a function that calls `_get_subtitles`, but only if subs are requested by the user. See common.py

File: yt_dlp/extractor/mgtv.py [hunk 9]
```diff
@@ -97,4 +129,26 @@ def _real_extract(self, url):
...
+                                   fatal=False) or {}
+        subtitles = {}
+        for sub in try_get(info, lambda x: x['data']['title']) or []:
+            url_sub = sub.get('url')
+            if not url_sub:
+                continue
+            locale = sub.get('captionCountrySimpleName')
+            sub = self._download_json(f'{domain}{url_sub}', video_id,
+                                      note=f'Download subtitle for locale {sub.get("name")} ({locale})',
+                                      fatal=False) or {}
```
Comment (pukkandan) [inline yt_dlp/extractor/mgtv.py:146]:
```suggestion
        info = self._download_json(f'https://pcweb.api.mgtv.com/video/title?videoId={video_id}',
                                   video_id, fatal=False) or {}
        subtitles = {}
        for sub in try_get(info, lambda x: x['data']['title']) or []:
            url_sub = sub.get('url')
            if not url_sub:
                continue
            locale = sub.get('captionCountrySimpleName')
            sub = self._download_json(f'{domain}{url_sub}', video_id, fatal=False,
                                      note=f'Download subtitle for locale {sub.get("name")} ({locale})') or {}
```

**Pull request merged**